driver-core/src/main/com/mongodb/connection/CommandProtocol.java

```diff
@@ -77,13 +83,40 @@ public boolean isSlaveOk() {
     public T execute(final InternalConnection connection) {
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug(format("Sending command {%s : %s} to database %s on connection [%s] to server %s",
-                                command.keySet().iterator().next(), command.values().iterator().next(),
+                                getCommandName(), command.values().iterator().next(),
                                 namespace.getDatabaseName(), connection.getDescription().getConnectionId(),
                                 connection.getDescription().getServerAddress()));
         }
-        T retval = receiveMessage(connection, sendMessage(connection).getId());
-        LOGGER.debug("Command execution completed");
-        return retval;
+        long startTimeNanos = System.nanoTime();
+        CommandMessage commandMessage = null;
+        try {
+            commandMessage = sendMessage(connection);
+            ResponseBuffers responseBuffers = connection.receiveMessage(commandMessage.getId());
+            ReplyMessage<BsonDocument> replyMessage;
+            try {
+                 replyMessage = new ReplyMessage<BsonDocument>(responseBuffers, new BsonDocumentCodec(), commandMessage.getId());
+            } finally {
+                responseBuffers.close();
+            }
+
+            BsonDocument response = replyMessage.getDocuments().get(0);
+            if (!ProtocolHelper.isCommandOk(response)) {
+                throw getCommandFailureException(response, connection.getDescription().getServerAddress());
+            }
+
+            T retval = commandResultDecoder.decode(new BsonDocumentReader(response), DecoderContext.builder().build());
+            if (commandListener != null) {
+                sendCommandSucceededEvent(commandMessage, getCommandName(), response, connection.getDescription(), startTimeNanos,
```

Comment:
startTimeNanos includes the time it took to raise to sendCommandStartedEvent event. Not sure where we ended up with that, but I believe I have the same thing. Should we be stopping the timer during the raising of the event? Technically, the time it takes to execute the command includes the time it takes to raise the event, so I can go both ways.